test/test-emacs-lisp-error-reporter.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright (C) 2020  Matthew "strager" Glazar
+// See end of file for extended copyright information.
+
+#include <cstring>
+#include <gtest/gtest.h>
+#include <quick-lint-js/char8.h>
+#include <quick-lint-js/emacs-lisp-error-reporter.h>
+#include <quick-lint-js/emacs-location.h>
+#include <quick-lint-js/location.h>
+#include <quick-lint-js/padded-string.h>
+#include <sstream>
+
+namespace quick_lint_js {
+namespace {
+class test_emacs_lisp_error_reporter : public ::testing::Test {
+ protected:
+  emacs_lisp_error_reporter make_reporter() {
+    return emacs_lisp_error_reporter(this->stream_);
+  }
+
+  emacs_lisp_error_reporter make_reporter(padded_string_view input) {
+    emacs_lisp_error_reporter reporter(this->stream_);
+    reporter.set_source(input);
+    return reporter;
+  }
+
+  std::string get_output() { return this->stream_.str(); }
+
+ private:
+  std::ostringstream stream_;
+};
+
+TEST_F(test_emacs_lisp_error_reporter, assignment_before_variable_declaration) {
+  padded_string input(u8"x=0;let x;"_sv);
+  source_code_span assignment_span(&input[1 - 1], &input[1 + 1 - 1]);
+  ASSERT_EQ(assignment_span.string_view(), u8"x");
+  source_code_span declaration_span(&input[9 - 1], &input[9 + 1 - 1]);
+  ASSERT_EQ(declaration_span.string_view(), u8"x");
+
+  emacs_lisp_error_reporter reporter = this->make_reporter(&input);
+  reporter.report(error_assignment_before_variable_declaration{
+      .assignment = identifier(assignment_span),
+      .declaration = identifier(declaration_span)});
+  reporter.finish();
+  EXPECT_EQ(
+      this->get_output(),
+      "(((1 . 2) 0 \"E001\" \"variable assigned before its declaration\"))");
+}
+
+TEST_F(test_emacs_lisp_error_reporter, assignment_to_const_global_variable) {
+  padded_string input(u8"to Infinity and beyond"_sv);
+  source_code_span infinity_span(&input[4 - 1], &input[11 + 1 - 1]);
+  ASSERT_EQ(infinity_span.string_view(), u8"Infinity");
+
+  emacs_lisp_error_reporter reporter = this->make_reporter(&input);
+
+  reporter.report(
+      error_assignment_to_const_global_variable{identifier(infinity_span)});
+  reporter.finish();
+  EXPECT_EQ(this->get_output(),
+            "(((4 . 12) 0 \"E002\" \"assignment to const global variable\"))");
+}
+
+TEST_F(test_emacs_lisp_error_reporter,
+       expected_parenthesis_around_if_condition) {
+  padded_string input(u8"if cond) {}"_sv);
+  source_code_span parenthesis_span(&input[4 - 1], &input[4 - 1]);
+  emacs_lisp_error_reporter reporter = this->make_reporter(&input);
+  reporter.report(error_expected_parenthesis_around_if_condition{
+      .where = parenthesis_span,
+      .token = '(',
+  });
+  reporter.finish();
+  EXPECT_EQ(this->get_output(),
+            "(((4 . 4) 0 \"E018\" \"if statement is missing '(' around "
+            "condition\"))");
```

Comment:
Nit: We can avoid backslashes by using a raw string literal:
```suggestion
            R"--((((4 . 4) 0 "E018" "if statement is missing '(' around condition")))--");
```
(But maybe this is more confusing because of the parentheses...)